File: android/src/main/java/com/gamepulse/sdk/GamePulse.java
package com.gamepulse.sdk;

import android.content.Context;
import android.os.Build;
import android.provider.Settings;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.WindowManager;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.TimeUnit;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class GamePulse {
    public enum Environment {
        DEVELOPMENT("https://client.dev.gamepulse.click/events/collect"),
        PRODUCTION("https://client.gamepulse.click/events/collect");
        
        private final String baseUrl;
        
        Environment(String baseUrl) {
            this.baseUrl = baseUrl;
        }
        
        public String getBaseUrl() {
            return baseUrl;
        }
    }
    
    // Platform is automatically detected as ANDROID for this SDK
    
    public static class UserConfig {
        private final String sessionId;
        private final String userId;
        private final String anonymousId;
        
        private UserConfig(String sessionId, String userId, String anonymousId) {
            if (sessionId == null || sessionId.isEmpty()) {
                throw new IllegalArgumentException("sessionId is mandatory");
            }
            if ((userId == null || userId.isEmpty()) && (anonymousId == null || anonymousId.isEmpty())) {
                throw new IllegalArgumentException("Either userId or anonymousId must be provided");
            }
            
            this.sessionId = sessionId;
            this.userId = userId;
            this.anonymousId = anonymousId;
        }
        
        public String getSessionId() { return sessionId; }
        public String getUserId() { return userId; }
        public String getAnonymousId() { return anonymousId; }
        
        public static Builder builder() {
            return new Builder();
        }
        
        public static class Builder {
            private String sessionId;
            private String userId;
            private String anonymousId;
            
            public Builder setSessionId(String sessionId) {
                this.sessionId = sessionId;
                return this;
            }
            
            public Builder setUserId(String userId) {
                this.userId = userId;
                return this;
            }
            
            public Builder setAnonymous(String anonymousId) {
                this.anonymousId = anonymousId;
                return this;
            }
            
            public UserConfig build() {
                return new UserConfig(sessionId, userId, anonymousId);
            }
        }
    }
    
    public static class DeviceInfo {
        private final String platform;
        private final String osVersion;
        private final String appVersion;
        private final String deviceModel;
        private final String screenResolution;
        private final String deviceManufacturer;
        
        public DeviceInfo(String platform, String osVersion, String appVersion, 
                         String deviceModel, String screenResolution, String deviceManufacturer) {
            this.platform = platform;
            this.osVersion = osVersion;
            this.appVersion = appVersion;
            this.deviceModel = deviceModel;
            this.screenResolution = screenResolution;
            this.deviceManufacturer = deviceManufacturer;
        }
        
        public String getPlatform() { return platform; }
        public String getOsVersion() { return osVersion; }
        public String getAppVersion() { return appVersion; }
        public String getDeviceModel() { return deviceModel; }
        public String getScreenResolution() { return screenResolution; }
        public String getDeviceManufacturer() { return deviceManufacturer; }
    }
    
    private static final MediaType JSON = MediaType.get("application/json; charset=utf-8");

    private static GameAlytics instance;
    private final Context context;
    private final String apiKey;
    private final Environment environment;
    private final boolean debug;
    private final OkHttpClient httpClient;
    
    private UserConfig userConfig;
    private DeviceInfo deviceInfo;
    private boolean isInitialized = false;

    // User Events
    public static class UserEvents {
        public static final String CATEGORY = "user";
        public static final String SESSION_START = "session_start";
        public static final String SESSION_END = "session_end";
        public static final String USER_LOGIN = "user_login";
        public static final String USER_LOGOUT = "user_logout";
        public static final String USER_REGISTER = "user_register";
        
        private static final String[] VALID_EVENTS = {
            SESSION_START, SESSION_END, USER_LOGIN, USER_LOGOUT, USER_REGISTER
        };
        
        public static boolean isValidEvent(String eventType) {
            for (String event : VALID_EVENTS) {
                if (event.equals(eventType)) return true;
            }
            return false;
        }
    }
    
    // Gameplay Events
    public static class GameplayEvents {
        public static final String CATEGORY = "gameplay";
        public static final String LEVEL_START = "level_start";
        public static final String LEVEL_END = "level_end";
        public static final String LEVEL_UP = "level_up";
        public static final String GAME_START = "game_start";
        public static final String GAME_END = "game_end";
        public static final String BOSS_FIGHT = "boss_fight";
        
        private static final String[] VALID_EVENTS = {
            LEVEL_START, LEVEL_END, LEVEL_UP, GAME_START, GAME_END, BOSS_FIGHT
        };
        
        public static boolean isValidEvent(String eventType) {
            for (String event : VALID_EVENTS) {
                if (event.equals(eventType)) return true;
            }
            return false;
        }
    }
    
    // Economy Events
    public static class EconomyEvents {
        public static final String CATEGORY = "economy";
        public static final String CURRENCY_EARNED = "currency_earned";
        public static final String CURRENCY_SPENT = "currency_spent";
        public static final String ITEM_PURCHASED = "item_purchased";
        public static final String ITEM_SOLD = "item_sold";
        public static final String SHOP_VIEWED = "shop_viewed";
        
        private static final String[] VALID_EVENTS = {
            CURRENCY_EARNED, CURRENCY_SPENT, ITEM_PURCHASED, ITEM_SOLD, SHOP_VIEWED
        };
        
        public static boolean isValidEvent(String eventType) {
            for (String event : VALID_EVENTS) {
                if (event.equals(eventType)) return true;
            }
            return false;
        }
    }
    
    // Progression Events
    public static class ProgressionEvents {
        public static final String CATEGORY = "progression";
        public static final String TUTORIAL_COMPLETE = "tutorial_complete";
        public static final String ACHIEVEMENT_UNLOCKED = "achievement_unlocked";
        public static final String MILESTONE_REACHED = "milestone_reached";
        public static final String QUEST_COMPLETED = "quest_completed";
        
        private static final String[] VALID_EVENTS = {
            TUTORIAL_COMPLETE, ACHIEVEMENT_UNLOCKED, MILESTONE_REACHED, QUEST_COMPLETED
        };
        
        public static boolean isValidEvent(String eventType) {
            for (String event : VALID_EVENTS) {
                if (event.equals(eventType)) return true;
            }
            return false;
        }
    }
    
    // Ad Events
    public static class AdEvents {
        public static final String CATEGORY = "ad";
        public static final String AD_VIEWED = "ad_viewed";
        public static final String AD_CLICKED = "ad_clicked";
        public static final String AD_REWARDED = "ad_rewarded";
        public static final String AD_FAILED = "ad_failed";
        
        private static final String[] VALID_EVENTS = {
            AD_VIEWED, AD_CLICKED, AD_REWARDED, AD_FAILED
        };
        
        public static boolean isValidEvent(String eventType) {
            for (String event : VALID_EVENTS) {
                if (event.equals(eventType)) return true;
            }
            return false;
        }
    }
    
    // IAP Events
    public static class IAPEvents {
        public static final String CATEGORY = "iap";
        public static final String PURCHASE = "purchase";
        public static final String PURCHASE_FAILED = "purchase_failed";
        public static final String PURCHASE_RESTORED = "purchase_restored";
        public static final String SUBSCRIPTION_STARTED = "subscription_started";
        public static final String SUBSCRIPTION_CANCELLED = "subscription_cancelled";
        
        private static final String[] VALID_EVENTS = {
            PURCHASE, PURCHASE_FAILED, PURCHASE_RESTORED, SUBSCRIPTION_STARTED, SUBSCRIPTION_CANCELLED
        };
        
        public static boolean isValidEvent(String eventType) {
            for (String event : VALID_EVENTS) {
                if (event.equals(eventType)) return true;
            }
            return false;
        }
    }

    private GamePulse(Context context, String apiKey, Environment environment) {
        this.context = context.getApplicationContext();
        this.apiKey = apiKey;
        this.environment = environment;
        this.debug = false;
        
        this.httpClient = new OkHttpClient.Builder()
                .connectTimeout(10, TimeUnit.SECONDS)
                .writeTimeout(10, TimeUnit.SECONDS)
                .readTimeout(30, TimeUnit.SECONDS)
                .build();
    }

    public static InitBuilder init(String apiKey, Environment environment) {
        return new InitBuilder(apiKey, environment);
    }

    public static GamePulse getInstance(Context context) {
        if (instance == null) {
            throw new IllegalStateException("GamePulse must be initialized first. Call GamePulse.init(...).create(context)");
        }
        return instance;
    }

    public static GamePulse getInstance() {
        if (instance == null) {
            throw new IllegalStateException("GamePulse must be initialized first. Call GamePulse.init(...).create(context)");
        }
        return instance;
    }

    public SystemEventBuilder systemEvent() {
        return new SystemEventBuilder();
    }

    public CustomEventBuilder customEvent() {
        return new CustomEventBuilder();
    }

    public static class InitBuilder {
        private final String apiKey;
        private final Environment environment;
        private UserConfig userConfig;
        
        private InitBuilder(String apiKey, Environment environment) {
            this.apiKey = apiKey;
            this.environment = environment;
        }
        
        public InitBuilder userConfig(UserConfig userConfig) {
            this.userConfig = userConfig;
            return this;
        }
        
        public GamePulse create(Context context) {
            if (apiKey == null || apiKey.isEmpty()) {
                throw new IllegalArgumentException("API key is required");
            }
            if (userConfig == null) {
                throw new IllegalArgumentException("UserConfig is required");
            }
            
            synchronized (GamePulse.class) {
                if (instance == null) {
                    instance = new GamePulse(context, apiKey, environment);
                    instance.userConfig = userConfig;
                    instance.deviceInfo = instance.autoFetchDeviceInfo();
                    instance.isInitialized = true;
                }
            }
            
            return instance;
        }
    }

    // Event Category Classes
    public static class Gameplay {
        public static final String LEVEL_START = "level_start";
        public static final String LEVEL_END = "level_end";
        public static final String LEVEL_UP = "level_up";
        public static final String BOSS_FIGHT = "boss_fight";
        public static final String CHECKPOINT_REACHED = "checkpoint_reached";
    }

    public static class IAP {
        public static final String PURCHASE = "purchase";
        public static final String PURCHASE_FAILED = "purchase_failed";
        public static final String PURCHASE_RESTORED = "purchase_restored";
        public static final String SUBSCRIPTION_STARTED = "subscription_started";
        public static final String SUBSCRIPTION_CANCELLED = "subscription_cancelled";
    }

    public static class User {
        public static final String SESSION_START = "session_start";
        public static final String SESSION_END = "session_end";
        public static final String USER_LOGIN = "user_login";
        public static final String USER_LOGOUT = "user_logout";
        public static final String USER_REGISTER = "user_register";
    }

    public static class Progression {
        public static final String TUTORIAL_COMPLETE = "tutorial_complete";
        public static final String ACHIEVEMENT_UNLOCKED = "achievement_unlocked";
        public static final String MILESTONE_REACHED = "milestone_reached";
        public static final String QUEST_COMPLETED = "quest_completed";
    }

    public static class Ad {
        public static final String AD_VIEWED = "ad_viewed";
        public static final String AD_CLICKED = "ad_clicked";
        public static final String AD_REWARDED = "ad_rewarded";
        public static final String AD_FAILED = "ad_failed";
    }

    // Event Builders
    public static class SystemEventBuilder {
        private String category;
        private String type;
        private Map<String, String> properties = new HashMap<>();

        public SystemEventBuilder category(Class<?> categoryClass) {
            this.category = categoryClass.getSimpleName().toLowerCase();
            return this;
        }

        public SystemEventBuilder type(String type) {
            this.type = type;
            return this;
        }

        public SystemEventBuilder setProperties(Map<String, String> properties) {
            this.properties = properties != null ? properties : new HashMap<>();
            return this;
        }

        public void trigger() {
            if (instance == null) {
                throw new IllegalStateException("GameAlytics must be initialized first");
            }
            if (category == null || type == null) {
                throw new IllegalArgumentException("Category and type are required");
            }
            instance.trackEventInternal("SYSTEM", type, category, properties);
        }
    }

    public static class CustomEventBuilder {
        private String category;
        private String type;
        private Map<String, String> properties = new HashMap<>();

        public CustomEventBuilder category(String category) {
            this.category = category;
            return this;
        }

        public CustomEventBuilder type(String type) {
            this.type = type;
            return this;
        }

        public CustomEventBuilder setProperties(Map<String, String> properties) {
            this.properties = properties != null ? properties : new HashMap<>();
            return this;
        }

        public void trigger() {
            if (instance == null) {
                throw new IllegalStateException("GameAlytics must be initialized first");
            }
            if (category == null || type == null) {
                throw new IllegalArgumentException("Category and type are required");
            }
            instance.trackEventInternal("CUSTOM", type, category, properties);
        }
    }

    private void trackEventInternal(String eventClass, String eventType, String category, Map<String, String> properties) {
        try {
            JSONObject json = new JSONObject();
            json.put("type", eventClass);
            json.put("value", eventType);
            json.put("category", category);
            
            // Add device info
            json.put("platform", deviceInfo.getPlatform());
            json.put("osVersion", deviceInfo.getOsVersion());
            json.put("deviceModel", deviceInfo.getDeviceModel());
            json.put("deviceManufacturer", deviceInfo.getDeviceManufacturer());
            json.put("appVersion", deviceInfo.getAppVersion());
            json.put("screenResolution", deviceInfo.getScreenResolution());
            
            // Add user/session info
            json.put("userId", userConfig.getUserId() != null ? userConfig.getUserId() : "");
            json.put("anonymousId", userConfig.getAnonymousId() != null ? userConfig.getAnonymousId() : "");
            json.put("sessionId", userConfig.getSessionId());
            
            // Add timestamp
            json.put("timezone", java.util.TimeZone.getDefault().getID());
            json.put("localDateTime", java.time.Instant.now().toString());
            
            // Add properties
            JSONObject propsJson = new JSONObject();
            for (Map.Entry<String, String> entry : properties.entrySet()) {
                propsJson.put(entry.getKey(), entry.getValue());
            }
            json.put("properties", propsJson);
            
            // Send the event
            sendEventInternal(json);
            
        } catch (JSONException e) {
            // Silent fail for performance
        }
    }
    
    public EventBuilder iapEvent(String eventType) {
        checkInitialized();
        if (!IAPEvents.isValidEvent(eventType)) {
            throw new IllegalArgumentException("Invalid IAP event type: " + eventType);
        }
        return new EventBuilder(eventType, IAPEvents.CATEGORY, false);
    }

    public EventBuilder customEvent(String type, String category) {
        checkInitialized();
        return new EventBuilder(type, category, true);
    }

    public void startSession() {
        checkInitialized();
        // Create new session in userConfig
        String newSessionId = UUID.randomUUID().toString();
        this.userConfig = new UserConfig(newSessionId, userConfig.getUserId(), userConfig.getAnonymousId());
        
        userEvent(UserEvents.SESSION_START)
            .setProperties(new HashMap<>())
            .track();
    }

    public void endSession() {
        checkInitialized();
        
        if (this.userConfig.getSessionId() == null) {
            return;
        }
        
        userEvent(UserEvents.SESSION_END)
            .setProperties(new HashMap<>())
            .track();
    }

    public void updateUserConfig(UserConfig newUserConfig) {
        checkInitialized();
        this.userConfig = newUserConfig;
    }
    
    public UserConfig getUserConfig() {
        return userConfig;
    }
    
    public DeviceInfo getDeviceInfo() {
        return deviceInfo;
    }
    
    // Missing event methods
    public EventBuilder userEvent(String eventType) {
        checkInitialized();
        if (!UserEvents.isValidEvent(eventType)) {
            throw new IllegalArgumentException("Invalid user event type: " + eventType);
        }
        return new EventBuilder(eventType, UserEvents.CATEGORY, false);
    }
    
    public EventBuilder gameplayEvent(String eventType) {
        checkInitialized();
        if (!GameplayEvents.isValidEvent(eventType)) {
            throw new IllegalArgumentException("Invalid gameplay event type: " + eventType);
        }
        return new EventBuilder(eventType, GameplayEvents.CATEGORY, false);
    }
    
    public EventBuilder adEvent(String eventType) {
        checkInitialized();
        if (!AdEvents.isValidEvent(eventType)) {
            throw new IllegalArgumentException("Invalid ad event type: " + eventType);
        }
        return new EventBuilder(eventType, AdEvents.CATEGORY, false);
    }
    
    public EventBuilder progressionEvent(String eventType) {
        checkInitialized();
        if (!ProgressionEvents.isValidEvent(eventType)) {
            throw new IllegalArgumentException("Invalid progression event type: " + eventType);
        }
        return new EventBuilder(eventType, ProgressionEvents.CATEGORY, false);
    }
    
    public EventBuilder economyEvent(String eventType) {
        checkInitialized();
        if (!EconomyEvents.isValidEvent(eventType)) {
            throw new IllegalArgumentException("Invalid economy event type: " + eventType);
        }
        return new EventBuilder(eventType, EconomyEvents.CATEGORY, false);
    }
    
    // Send event internally
    private void sendEventInternal(JSONObject eventData) {
        RequestBody body = RequestBody.create(eventData.toString(), JSON);
        Request request = new Request.Builder()
                .url(environment.getBaseUrl())
                .addHeader("Content-Type", "application/json")
                .addHeader("x-api-key", apiKey)
                .post(body)
                .build();

        httpClient.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) {
                if (debug) {
                    System.out.println("GameAlytics: Failed to send event: " + e.getMessage());
                }
            }

            @Override
            public void onResponse(Call call, Response response) throws IOException {
                if (debug) {
                    System.out.println("GameAlytics: Event sent successfully: " + response.code());
                }
                response.close();
            }
        });
    }
    
    // Public sendEvent method (for backwards compatibility)
    private void sendEvent(JSONObject eventData) {
        sendEventInternal(eventData);
    }
    

    public class EventBuilder {
        private final String eventType;
        private final String eventCategory;
        private final boolean isCustom;
        private Map<String, String> properties = new HashMap<>();


        EventBuilder(String type, String category, boolean isCustom) {
            this.eventType = type;
            this.eventCategory = category;
            this.isCustom = isCustom;
        }

        public EventBuilder setProperties(Map<String, String> properties) {
            if (properties != null) {
                this.properties = properties;
            }
            return this;
        }

        public void track() {
            checkInitialized();
            
            JSONObject json = new JSONObject();
            try {
                // Add common fields
                json.put("type", isCustom ? "CUSTOM" : "SYSTEM");
                json.put("value", eventType);
                json.put("category", eventCategory);
                
                // Add device info
                json.put("platform", deviceInfo.getPlatform());
                json.put("osVersion", deviceInfo.getOsVersion());
                json.put("deviceModel", deviceInfo.getDeviceModel());
                json.put("deviceManufacturer", deviceInfo.getDeviceManufacturer());
                json.put("appVersion", deviceInfo.getAppVersion());
                json.put("screenResolution", deviceInfo.getScreenResolution());
                
                // Add user/session info
                json.put("userId", userConfig.getUserId() != null ? userConfig.getUserId() : "");
                json.put("anonymousId", userConfig.getAnonymousId() != null ? userConfig.getAnonymousId() : "");
                json.put("sessionId", userConfig.getSessionId());
                
                // Add timestamp
                json.put("timezone", java.util.TimeZone.getDefault().getID());
                json.put("localDateTime", java.time.Instant.now().toString());
                
                // Add properties
                JSONObject propsJson = new JSONObject();
                for (Map.Entry<String, String> entry : properties.entrySet()) {
                    propsJson.put(entry.getKey(), entry.getValue());
                }
                json.put("properties", propsJson);
                
                // Send the event
                GamePulse.this.sendEventInternal(json);
                
            } catch (JSONException e) {
                // Silent fail for performance
            }
        }
    }

    private DeviceInfo autoFetchDeviceInfo() {
        try {
            String appVersion = getAppVersion();
            String osVersion = Build.VERSION.RELEASE;
            String deviceModel = Build.MODEL;
            String deviceManufacturer = Build.MANUFACTURER;
            String screenResolution = getScreenResolution();
            
            return new DeviceInfo(
                "ANDROID", // Platform is hardcoded for Android SDK
                osVersion,
                appVersion,
                deviceModel,
                screenResolution,
                deviceManufacturer
            );
        } catch (Exception e) {
            // Silent fail for performance
            // Return default values if auto-fetch fails
            return new DeviceInfo(
                "ANDROID", // Platform is hardcoded for Android SDK
                "unknown",
                "1.0.0",
                "unknown",
                "unknown",
                "unknown"
            );
        }
    }
    
    private String getAppVersion() {
        try {
            return context.getPackageManager()
                    .getPackageInfo(context.getPackageName(), 0)
                    .versionName;
        } catch (Exception e) {
            return "1.0.0";
        }
    }

    private String getScreenResolution() {
        try {
            WindowManager windowManager = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
            if (windowManager != null) {
                DisplayMetrics displayMetrics = new DisplayMetrics();
                windowManager.getDefaultDisplay().getMetrics(displayMetrics);
                return displayMetrics.widthPixels + "x" + displayMetrics.heightPixels;
            }
        } catch (Exception e) {
            // Silent fail for performance
        }
        return "unknown";
    }

    private void checkInitialized() {
        if (!isInitialized) {
            throw new IllegalStateException("GameAlytics must be initialized first. Call GameAlytics.init()");
        }
    }
}

File: web/src/GamePulse.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Environment = exports.UserConfig = exports.GameAlytics = exports.CustomEventBuilder = exports.SystemEventBuilder = exports.Ad = exports.Progression = exports.User = exports.IAP = exports.Gameplay = void 0;
// Core types
var DataValidator_1 = require("./validation/DataValidator");
var EventQueue_1 = require("./queue/EventQueue");
var Environment;
(function (Environment) {
    Environment["DEVELOPMENT"] = "DEVELOPMENT";
    Environment["PRODUCTION"] = "PRODUCTION";
})(Environment || (exports.Environment = Environment = {}));
var UserConfigBuilder = /** @class */ (function () {
    function UserConfigBuilder() {
    }
    UserConfigBuilder.prototype.setSessionId = function (sessionId) {
        this.sessionId = sessionId;
        return this;
    };
    UserConfigBuilder.prototype.setUserId = function (userId) {
        this.userId = userId;
        return this;
    };
    UserConfigBuilder.prototype.setAnonymous = function (anonymousId) {
        this.anonymousId = anonymousId;
        return this;
    };
    UserConfigBuilder.prototype.build = function () {
        if (!this.sessionId) {
            throw new Error('SessionId is required');
        }
        if (!this.userId && !this.anonymousId) {
            throw new Error('Either userId or anonymousId must be provided');
        }
        return new UserConfig(this.sessionId, this.userId, this.anonymousId);
    };
    return UserConfigBuilder;
}());
var UserConfig = /** @class */ (function () {
    function UserConfig(sessionId, userId, anonymousId) {
        this.sessionId = sessionId;
        if (userId !== undefined) {
            this.userId = userId;
        }
        if (anonymousId !== undefined) {
            this.anonymousId = anonymousId;
        }
    }
    UserConfig.builder = function () {
        return new UserConfigBuilder();
    };
    return UserConfig;
}());
exports.UserConfig = UserConfig;
var EventBuilderImpl = /** @class */ (function () {
    function EventBuilderImpl(eventType, eventCategory, isCustom) {
        this.properties = {};
        this.eventType = eventType;
        this.eventCategory = eventCategory;
        this.isCustom = isCustom;
    }
    EventBuilderImpl.prototype.setProperties = function (properties) {
        this.properties = __assign(__assign({}, this.properties), properties);
        return this;
    };
    EventBuilderImpl.prototype.track = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                GameAlytics.getInstance().trackEvent(this.isCustom ? 'CUSTOM' : 'SYSTEM', this.eventType, this.eventCategory, this.properties);
                return [2 /*return*/];
            });
        });
    };
    return EventBuilderImpl;
}());
var InitBuilder = /** @class */ (function () {
    function InitBuilder(apiKey, environment) {
        this.apiKey = apiKey;
        this.environment = environment;
    }
    InitBuilder.prototype.userConfig = function (userConfig) {
        this.userConfigValue = userConfig;
        return this;
    };
    InitBuilder.prototype.create = function () {
        if (!this.apiKey) {
            throw new Error('API key is required');
        }
        if (!this.userConfigValue) {
            throw new Error('UserConfig is required');
        }
        GamePulse.instance = new GamePulse(this.apiKey, this.environment, this.userConfigValue);
        return GamePulse.instance;
    };
    return InitBuilder;
}());
var GamePulse = /** @class */ (function () {
    function GamePulse(apiKey, environment, userConfig) {
        this.isInitialized = false;
        this.apiKey = apiKey;
        this.environment = environment;
        this.userConfig = userConfig;
        this.deviceInfo = this.autoFetchDeviceInfo();
        this.baseUrl = environment === 'production' ? 'https://client.gamepulse.click' : 'https://client.dev.gamepulse.click';
        // Initialize event queue with base URL and API key injection
        this.eventQueue = new EventQueue_1.EventQueue();
        window.__GA_BASE_URL__ = this.baseUrl;
        window.__GA_API_KEY__ = this.apiKey;
        this.isInitialized = true;
    }
    GamePulse.init = function (apiKey, environment) {
        return new InitBuilder(apiKey, environment);
    };
    GamePulse.getInstance = function () {
        if (!GamePulse.instance) {
            throw new Error('GamePulse must be initialized first. Call GamePulse.init(...).create()');
        }
        return GamePulse.instance;
    };
    GamePulse.prototype.systemEvent = function () {
        return new SystemEventBuilder();
    };
    GamePulse.prototype.customEvent = function () {
        return new CustomEventBuilder();
    };
    GamePulse.prototype.autoFetchDeviceInfo = function () {
        try {
            var isBrowser = typeof window !== 'undefined';
            return {
                platform: isBrowser ? 'Web' : 'Node.js',
                osVersion: this.getOSVersion(),
                deviceModel: isBrowser ? window.navigator.platform : 'Node.js',
                screenResolution: this.getScreenResolution(),
                deviceManufacturer: isBrowser ? window.navigator.vendor || 'Unknown' : 'Node.js',
                appVersion: '1.0.0'
            };
        }
        catch (error) {
            // Silent failure for performance
            return {
                platform: 'Unknown',
                osVersion: 'Unknown',
                deviceModel: 'Unknown',
                screenResolution: 'Unknown',
                deviceManufacturer: 'Unknown',
                appVersion: '1.0.0'
            };
        }
    };
    GameAlytics.prototype.trackEvent = function (type, eventType, category, properties) {
        // Validate and sanitize properties
        var validationResult = DataValidator_1.DataValidator.validateProperties(properties, {
            maxLength: 100,
            allowHtml: false,
            allowedTypes: ['string', 'number', 'boolean']
        });
        // Log validation errors in development
        if (validationResult.errors.length > 0 && this.environment === Environment.DEVELOPMENT) {
            console.warn('GameAlytics validation warnings:', validationResult.errors);
        }
        var eventPayload = __assign(__assign({ type: type, value: eventType, category: category, properties: validationResult.sanitized }, this.deviceInfo), { timezone: Intl.DateTimeFormat().resolvedOptions().timeZone, localDateTime: new Date().toISOString(), userId: this.userConfig.userId || '', anonymousId: this.userConfig.anonymousId || '', sessionId: this.userConfig.sessionId });
        // Use event queue for reliable delivery
        this.eventQueue.enqueue(eventPayload);
    };
    GamePulse.prototype.getOSVersion = function () {
        if (typeof window === 'undefined')
            return 'Node.js';
        var userAgent = window.navigator.userAgent;
        var platform = window.navigator.platform;
        var macosPlatforms = ['Macintosh', 'MacIntel', 'MacPPC', 'Mac68K'];
        var windowsPlatforms = ['Win32', 'Win64', 'Windows', 'WinCE'];
        var iosPlatforms = ['iPhone', 'iPad', 'iPod'];
        if (macosPlatforms.indexOf(platform) !== -1)
            return 'Mac OS';
        if (iosPlatforms.indexOf(platform) !== -1)
            return 'iOS';
        if (windowsPlatforms.indexOf(platform) !== -1)
            return 'Windows';
        if (/Android/.test(userAgent))
            return 'Android';
        if (/Linux/.test(platform))
            return 'Linux';
        return 'Unknown';
    };
    GamePulse.prototype.getScreenResolution = function () {
        if (typeof window !== 'undefined' && window.screen) {
            return "".concat(window.screen.width, "x").concat(window.screen.height);
        }
        return 'unknown';
    };
    GamePulse.prototype.ensureInitialized = function () {
        if (!this.isInitialized) {
            throw new Error('GamePulse must be initialized first. Call GamePulse.init()');
        }
    };
    GamePulse.instance = null;
    return GamePulse;
}());
exports.GamePulse = GamePulse;
// Event Category Classes
var Gameplay = /** @class */ (function () {
    function Gameplay() {
    }
    Gameplay.LEVEL_START = 'level_start';
    Gameplay.LEVEL_END = 'level_end';
    Gameplay.LEVEL_UP = 'level_up';
    Gameplay.BOSS_FIGHT = 'boss_fight';
    Gameplay.CHECKPOINT_REACHED = 'checkpoint_reached';
    return Gameplay;
}());
exports.Gameplay = Gameplay;
var IAP = /** @class */ (function () {
    function IAP() {
    }
    IAP.PURCHASE = 'purchase';
    IAP.PURCHASE_FAILED = 'purchase_failed';
    IAP.PURCHASE_RESTORED = 'purchase_restored';
    IAP.SUBSCRIPTION_STARTED = 'subscription_started';
    IAP.SUBSCRIPTION_CANCELLED = 'subscription_cancelled';
    return IAP;
}());
exports.IAP = IAP;
var User = /** @class */ (function () {
    function User() {
    }
    User.SESSION_START = 'session_start';
    User.SESSION_END = 'session_end';
    User.USER_LOGIN = 'user_login';
    User.USER_LOGOUT = 'user_logout';
    User.USER_REGISTER = 'user_register';
    return User;
}());
exports.User = User;
var Progression = /** @class */ (function () {
    function Progression() {
    }
    Progression.TUTORIAL_COMPLETE = 'tutorial_complete';
    Progression.ACHIEVEMENT_UNLOCKED = 'achievement_unlocked';
    Progression.MILESTONE_REACHED = 'milestone_reached';
    Progression.QUEST_COMPLETED = 'quest_completed';
    return Progression;
}());
exports.Progression = Progression;
var Ad = /** @class */ (function () {
    function Ad() {
    }
    Ad.AD_VIEWED = 'ad_viewed';
    Ad.AD_CLICKED = 'ad_clicked';
    Ad.AD_REWARDED = 'ad_rewarded';
    Ad.AD_FAILED = 'ad_failed';
    return Ad;
}());
exports.Ad = Ad;
// Event Builders
var SystemEventBuilder = /** @class */ (function () {
    function SystemEventBuilder() {
        this.properties = {};
    }
    SystemEventBuilder.prototype.categoryClass = function (categoryClass) {
        this.category = categoryClass.constructor.name.toLowerCase();
        return this;
    };
    SystemEventBuilder.prototype.eventType = function (type) {
        this.type = type;
        return this;
    };
    SystemEventBuilder.prototype.setProperties = function (properties) {
        this.properties = properties || {};
        return this;
    };
    SystemEventBuilder.prototype.trigger = function () {
        if (!GameAlytics.instance) {
            throw new Error('GameAlytics must be initialized first');
        }
        if (!this.category || !this.type) {
            throw new Error('Category and type are required');
        }
        GameAlytics.instance.trackEvent('SYSTEM', this.type, this.category, this.properties);
    };
    return SystemEventBuilder;
}());
exports.SystemEventBuilder = SystemEventBuilder;
var CustomEventBuilder = /** @class */ (function () {
    function CustomEventBuilder() {
        this.properties = {};
    }
    CustomEventBuilder.prototype.categoryName = function (category) {
        this.category = category;
        return this;
    };
    CustomEventBuilder.prototype.eventType = function (type) {
        this.type = type;
        return this;
    };
    CustomEventBuilder.prototype.setProperties = function (properties) {
        this.properties = properties || {};
        return this;
    };
    CustomEventBuilder.prototype.trigger = function () {
        if (!GameAlytics.instance) {
            throw new Error('GameAlytics must be initialized first');
        }
        if (!this.category || !this.type) {
            throw new Error('Category and type are required');
        }
        GameAlytics.instance.trackEvent('CUSTOM', this.type, this.category, this.properties);
    };
    return CustomEventBuilder;
}());
exports.CustomEventBuilder = CustomEventBuilder;
exports.default = GameAlytics;
